lib/opam_solve.ml

```diff
@@ -251,27 +335,47 @@ module Make_solver (Context : OPAM_MONOREPO_CONTEXT) :
       rolemap []
     |> List.filter_map ~f:Solver.package_name
 
-  let get_opam_info ~context pkg =
-    match Context.opam_file context pkg with
-    | Ok opam_file -> Opam.Package_summary.from_opam ~pkg opam_file
+  let get_opam_info ~context { package; vendored } =
+    match Context.opam_file context package with
+    | Ok opam_file -> Opam.Package_summary.from_opam ~vendored package opam_file
     | Error (`Msg msg) ->
         (* If we're calling this function on a package, it means it has been
            returned as part of the solver solution and therefore should correspond
            to a valid candidate. *)
         Logs.debug (fun l ->
-            l "Could not retrieve opam file for %a: %s" Opam.Pp.package pkg msg);
+            l "Could not retrieve opam file for %a: %s" Opam.Pp.package package
+              msg);
         assert false
 
-  (* TODO catch exceptions and turn to error *)
-
   let calculate ~build_only ~allow_jbuilder ~local_opam_files:local_packages
       ~target_packages ~pin_depends ?ocaml_version input =
     let open Result.O in
+    let* opam_context =
+      build_context ~build_only ~allow_jbuilder ~ocaml_version ~pin_depends
+        ~local_packages ~target_packages ~require_dune:false input
+    in
+    let opam_provided =
+      match opam_provided_packages local_packages target_packages with
+      | Ok opam_provided -> opam_provided
+      | Error msg ->
+          Logs.warn (fun l ->
+              l "Error parsing x-opam-monorepo-provided: %s, ignoring.\n" msg);
+          OpamPackage.Name.Set.empty
+    in
+
+    let allow_compiler_variants = depend_on_compiler_variants local_packages in
+    let* opam_provided_deps =
+      determine_dependencies opam_context
```

Comment:
I don't think the opam-provided packages should be solved first.

The problem with this approach is that we don't have the insurance that the opam provided and duniverse packages require the same ocaml or dune versions. It also reverses things, opam provided packages deps prime over duniverse ones and that shouldn't be the case. Assembling a valid duniverse should be the priority here. Adding an opam provided package can strip out a dependency from the duniverse here.

One thing we can eventually try is a single solver run, where we dynamically add to a list of dune exempt packages. Happy to discuss this further offline!

Reply:
As we have discussed offline I have implemented as follows:

1. Filter out all the opam-provided packages from all the formulas that we feed into the solver
2. Run the solver, and get a list of packages + versions. These are supposed to be dune-installable
3. Create a context (`opam_context`) where we packages are allowed not to use dune, inject a dummy package (`opam-monorepo-all-dependencies.1`) into it. The solver does only allow solving for `OpamPackage.Name.t` not `OpamPackage.t`, hence the dummy package which depends on everything from step 2 + opam-provided packages
4. Run the solver again, thus getting a solution with both our vendored packages at exactly the versions we wanted as well as the additional opam packages at versions that are co-installable with the rest